File: data/models/RequestsCache.js
// models/RequestsCache.js
const mongoose = require( 'mongoose' );
const BaseSchema = require( './BaseSchema' );

const ExcludedSchema = new mongoose.Schema( {
  mbid:            { type: String, required: true },
  reason_excluded: { type: String },
  date_excluded:   { type: Date, default: Date.now },
} );

const RequestsCacheSchema = new mongoose.Schema( {
  params: {
    language:    String,
    genre:       String,
    niche_level: String,
  },
  excluded: [ ExcludedSchema ],
} );

RequestsCacheSchema.add( BaseSchema );

module.exports = mongoose.model( 'RequestsCache', RequestsCacheSchema );

File: backend/src/models/RequestsCache.py
# requestsCache.py

from mongoengine import (
    EmbeddedDocument,
    EmbeddedDocumentField,
    ListField,
    StringField,
    DateTimeField
)
from datetime import datetime, timezone
from BaseSchema import BaseSchema

class Excluded(EmbeddedDocument):
    mbid            = StringField(required=True)
    reason_excluded = StringField(required=True)
    date_excluded   = DateTimeField(default=datetime.now(timezone.utc), required=True)

class ParamsCache(EmbeddedDocument):
    language    = StringField(required=True)
    genre       = StringField(required=True)
    niche_level = StringField(required=True)

class RequestsCache(BaseSchema):
    params   = EmbeddedDocumentField(ParamsCache)
    excluded = ListField(EmbeddedDocumentField(Excluded))

    meta = {
        'collection': 'requests',
        'indexes'   : [
            {'fields': ['params.language']},
            {'fields': ['params.genre']},
            {'fields': ['params.niche_level']}
        ]
    }

